Gracefully handle symbolic links pointing to non-existing files in output directories

Original issue: https://issues.gradle.org/browse/GRADLE-1843

**Highly voted issue: 21**

### Expected Behavior

Ignore dangling symlinks when calculating the outputs for UP-TO-DATE checking. Potentially introduce a property that allows for configuring if a symlink should be followed.

### Current Behavior

Gradle fails with an exception when a dangling symlink is found.

### Context

Gradle should handle dangling symlinks more gracefully.

### Steps to Reproduce (for bugs)

Given the following tasks:

```
task testCopy(type: Copy) {
    from "test"
    into "$buildDir/test"
}

task testSync(type: Sync) {
    from "test"
    into "$buildDir/test"
}
```

And the following file structure:

```
mkdir -p build/test
[:~/tmp/test] % ln -s ../foo build/test
[:~/tmp/test] % touch build/test/valid
[:~/tmp/test] % ln -s valid build/test/valid_link
[:~/tmp/test] % ls -l build/test test
build/test:
total 0
-rw-r--r-- 1 alpar users 0 Oct 12 11:44 file.txt
lrwxrwxrwx 1 alpar users 6 Oct 12 11:46 foo -> ../foo
-rw-r--r-- 1 alpar users 0 Oct 12 11:46 valid
lrwxrwxrwx 1 alpar users 5 Oct 12 11:46 valid_link -> valid

test:
total 0
-rw-r--r-- 1 alpar users 0 Oct 12 09:18 file.txt
```

The following exception is thrown:

```
[:~/tmp/test] 1 % ./gradlew -s testCopy
* Exception is:
org.gradle.api.GradleException: Could not list contents of '/tmp/alpar/test/build/test/foo'. Couldn't follow symbolic link.
	at org.gradle.api.internal.file.collections.jdk7.Jdk7DirectoryWalker$1.visitFile(Jdk7DirectoryWalker.java:79)
	at org.gradle.api.internal.file.collections.jdk7.Jdk7DirectoryWalker$1.visitFile(Jdk7DirectoryWalker.java:54)
	at org.gradle.api.internal.file.collections.jdk7.Jdk7DirectoryWalker.walkDir(Jdk7DirectoryWalker.java:54)
	at org.gradle.api.internal.file.collections.DirectoryFileTree.walkDir(DirectoryFileTree.java:146)
	at org.gradle.api.internal.file.collections.DirectoryFileTree.visitFrom(DirectoryFileTree.java:130)
	at org.gradle.api.internal.file.collections.DirectoryFileTree.visit(DirectoryFileTree.java:115)
	at org.gradle.api.internal.file.collections.DirectoryFileTree.visitTreeOrBackingFile(DirectoryFileTree.java:111)
	at org.gradle.api.internal.file.collections.FileTreeAdapter.visitTreeOrBackingFile(FileTreeAdapter.java:114)
	at org.gradle.api.internal.changedetection.state.DefaultFileCollectionSnapshotter.visitFiles(DefaultFileCollectionSnapshotter.java:42)
	at org.gradle.api.internal.changedetection.state.AbstractFileCollectionSnapshotter.snapshot(AbstractFileCollectionSnapshotter.java:62)
	at org.gradle.api.internal.changedetection.state.DefaultFileCollectionSnapshotter.snapshot(DefaultFileCollectionSnapshotter.java:31)
	at org.gradle.api.internal.changedetection.state.AbstractFileCollectionSnapshotter.snapshot(AbstractFileCollectionSnapshotter.java:100)
	at org.gradle.api.internal.changedetection.state.DefaultFileCollectionSnapshotter.snapshot(DefaultFileCollectionSnapshotter.java:31)
	at org.gradle.api.internal.changedetection.state.OutputFilesCollectionSnapshotter.snapshot(OutputFilesCollectionSnapshotter.java:60)
	at org.gradle.api.internal.changedetection.rules.AbstractNamedFileSnapshotTaskStateChanges.buildSnapshots(AbstractNamedFileSnapshotTaskStateChanges.java:85)
	at org.gradle.api.internal.changedetection.rules.AbstractNamedFileSnapshotTaskStateChanges.<init>(AbstractNamedFileSnapshotTaskStateChanges.java:53)
	at org.gradle.api.internal.changedetection.rules.OutputFilesTaskStateChanges.<init>(OutputFilesTaskStateChanges.java:31)
	at org.gradle.api.internal.changedetection.rules.TaskUpToDateState.<init>(TaskUpToDateState.java:54)
	at org.gradle.api.internal.changedetection.changes.DefaultTaskArtifactStateRepository$TaskArtifactStateImpl.getStates(DefaultTaskArtifactStateRepository.java:162)
	at org.gradle.api.internal.changedetection.changes.DefaultTaskArtifactStateRepository$TaskArtifactStateImpl.isUpToDate(DefaultTaskArtifactStateRepository.java:82)
	at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:52)
	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
	at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:52)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:53)
	at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:233)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:215)
	at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.processTask(AbstractTaskPlanExecutor.java:74)
	at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.run(AbstractTaskPlanExecutor.java:55)
	at org.gradle.execution.taskgraph.ParallelTaskPlanExecutor.process(ParallelTaskPlanExecutor.java:50)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.execute(DefaultTaskGraphExecuter.java:113)
	at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:37)
	at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
	at org.gradle.execution.DefaultBuildExecuter.access$000(DefaultBuildExecuter.java:23)
	at org.gradle.execution.DefaultBuildExecuter$1.proceed(DefaultBuildExecuter.java:43)
	at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:32)
	at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
	at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:30)
	at org.gradle.initialization.DefaultGradleLauncher$4.run(DefaultGradleLauncher.java:186)
	at org.gradle.internal.Factories$1.create(Factories.java:22)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:91)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:53)
	at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:183)
	at org.gradle.initialization.DefaultGradleLauncher.access$200(DefaultGradleLauncher.java:33)
	at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:112)
	at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:106)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:91)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:63)
	at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:106)
	at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:92)
	at org.gradle.launcher.exec.GradleBuildController.run(GradleBuildController.java:66)
	at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:28)
	at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:41)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:26)
	at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:79)
	at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:51)
	at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:59)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:47)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:26)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:34)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:74)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:72)
	at org.gradle.util.Swapper.swap(Swapper.java:38)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:72)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:55)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:60)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:72)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.HintGCAfterBuild.execute(HintGCAfterBuild.java:44)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:50)
	at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:293)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
```

I think the real problem is that the Copy task comes with an **incorrect UP-TO-DATE check configuration**. I'm using Gradle 4.3 and the following is my test project:

```groovy
task generateConfigFile(type: Copy) {
	from '.'
	into '.'
	include 'config.file.template'
	rename { fileName -> 'config.file' }
	expand a: 'b'
}

task clean(type: Delete) {
	delete 'config.file'
}

// =============
// For debugging
// =============
private void printTaskDetails(Task task) {
	StringBuffer taskDetails = new StringBuffer()
	taskDetails << """-------------
name: $task.name group:$task.group : $task.description
conv: $task.convention.plugins
inputs:
"""
	task.inputs.files.each { it ->
		taskDetails << "\t${it.absolutePath}\n"
	}
	taskDetails << "outputs:\n"
	task.outputs.files.each { it ->
		taskDetails << "\t${it.absolutePath}\n"
	}
	taskDetails << "-------------\n"
	println taskDetails
}

gradle.taskGraph.afterTask { task ->
	printTaskDetails task
}
```

Then I execute the following commands:

```none
$ gw generateConfigFile
(...)
> Task :generateConfigFile 
-------------
name: generateConfigFile group:null : null
conv: [:]
inputs:
        /home/rkrisztian/projects/gradletest/config.file.template
outputs:
        /home/rkrisztian/projects/gradletest
-------------



BUILD SUCCESSFUL in 4s
1 actionable task: 1 executed

$ ln -s /a/b build/broken.symlink

$ gw generateConfigFile
(...)
> Task :generateConfigFile FAILED
-------------
name: generateConfigFile group:null : null
conv: [:]
inputs:
        /home/rkrisztian/projects/gradletest/config.file.template
outputs:
        /home/rkrisztian/projects/gradletest
-------------



FAILURE: Build failed with an exception.

* What went wrong:
Failed to capture snapshot of output files for task ':generateConfigFile' property 'destinationDir' during up-to-date check.
> Could not list contents of '/home/rkrisztian/projects/gradletest/build/broken.symlink'. Couldn't follow symbolic link.
(...)
```

I think it is just plain wrong to define the whole directory as an output, when in fact it should be a single file. Please correct me if I'm wrong.

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy
@@ -18,15 +18,16 @@ package org.gradle.api.file
 import org.gradle.integtests.fixtures.AbstractIntegrationSpec
 import org.gradle.util.Requires
 import org.gradle.util.TestPrecondition
+import spock.lang.Issue
 import spock.lang.Unroll
 
 import static org.gradle.util.TextUtil.escapeString
 
+@Unroll
 @Requires(TestPrecondition.SYMLINKS)
 class FileCollectionSymlinkIntegrationTest extends AbstractIntegrationSpec {
 
-    @Unroll("#desc can handle symlinks")
-    def "file collection can handle symlinks"() {
+    def "#desc can handle symlinks"() {
         def buildScript = file("build.gradle")
         def baseDir = file('build')
         baseDir.file('file').text = 'some contents'
@@ -63,6 +64,28 @@ class FileCollectionSymlinkIntegrationTest extends AbstractIntegrationSpec {
         "project.layout.files()"             | "project.layout.files(file, symlink, symlinked)"
         "project.layout.configurableFiles()" | "project.layout.configurableFiles(file, symlink, symlinked)"
         "project.objects.fileCollection()"   | "project.objects.fileCollection().from(file, symlink, symlinked)"
+    }
+
+    @Issue('https://github.com/gradle/gradle/issues/1365')
+    def "broken symlink not produced by task is ignored"() {
+        given:
+        def input = file("input.txt").createFile()
+        def outputDirectory = file("output")
+
+        def brokenLink = file('link').createLink("broken")
+        assert !brokenLink.exists()
+
+        buildFile << """
+            task copy(type: Copy) {
+                from '${input.name}'
+                into '${outputDirectory.name}'
+            }
+        """
+
+        when:
+        succeeds 'copy'
+        then:
+        outputDirectory.list().contains input.name
     }
 
     void maybeDeprecated(String expression) {

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy
@@ -155,9 +155,18 @@ class FileCollectionSymlinkIntegrationTest extends AbstractIntegrationSpec {
         """
 
         when:
-        succeeds 'copy'
+        run 'copy'
         then:
+        executedAndNotSkipped ':copy'
         outputDirectory.list().contains input.name
+
+        when:
+        run 'copy'
+        then:
+        skipped ':copy'
+
+        cleanup:
+        brokenLink.delete()
     }
 
     void maybeDeprecated(String expression) {

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy
@@ -66,6 +66,52 @@ class FileCollectionSymlinkIntegrationTest extends AbstractIntegrationSpec {
         "project.objects.fileCollection()"   | "project.objects.fileCollection().from(file, symlink, symlinked)"
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/1365")
+    def "detect changes to broken symlink outputs"() {
+        def root = file("root").createDir()
+        def target = file("target")
+        def link = root.file("link")
+
+        buildFile << """
+            import java.nio.file.*
+            class ProducesLink extends DefaultTask {
+                @OutputDirectory File outputDirectory // TODO outputfile?
+    
+                @TaskAction execute() {
+                    def link = Paths.get('${link}')
+                    Files.deleteIfExists(link);
+                    Files.createSymbolicLink(link, Paths.get('${target}'));
+                }
+            }
+            
+            task producesLink(type: ProducesLink) {
+                outputDirectory = file '${root.name}'
+            }
+        """
+
+        when:
+        target.createFile()
+        run 'producesLink'
+        then:
+        executedAndNotSkipped ':producesLink'
+
+        when:
+        run 'producesLink'
+        then:
+        skipped ':producesLink'
+
+        when:
+        target.delete()
+        run 'producesLink'
+        then:
+        executedAndNotSkipped ':producesLink'
+
+        when:
+        run 'producesLink'
+        then:
+        skipped ':producesLink'
+    }
+
     @Issue('https://github.com/gradle/gradle/issues/1365')
     def "broken symlink not produced by task is ignored"() {
         given:

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy
@@ -158,15 +158,20 @@ class FileCollectionSymlinkIntegrationTest extends AbstractIntegrationSpec {
         run 'copy'
         then:
         executedAndNotSkipped ':copy'
-        outputDirectory.list().contains input.name
+        outputDirectory.list() == [input.name, brokenLink.name]
 
         when:
         run 'copy'
         then:
         skipped ':copy'
+        outputDirectory.list() == [input.name, brokenLink.name]
 
-        cleanup:
+        when:
         brokenLink.delete()
+        run 'copy'
+        then:
+        skipped ':copy'
+        outputDirectory.list() == [input.name]
     }
 
     void maybeDeprecated(String expression) {

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy
@@ -118,7 +118,7 @@ class FileCollectionSymlinkIntegrationTest extends AbstractIntegrationSpec {
         def input = file("input.txt").createFile()
         def outputDirectory = file("output")
 
-        def brokenLink = file('link').createLink("broken")
+        def brokenLink = outputDirectory.file('link').createLink("broken")
         assert !brokenLink.exists()
 
         buildFile << """

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy
@@ -67,27 +67,11 @@ class FileCollectionSymlinkIntegrationTest extends AbstractIntegrationSpec {
     }
 
     @Issue("https://github.com/gradle/gradle/issues/1365")
-    def "detect changes to broken symlink outputs"() {
+    def "detect changes to broken symlink outputs in #outputType"() {
         def root = file("root").createDir()
         def target = file("target")
         def link = root.file("link")
-
-        buildFile << """
-            import java.nio.file.*
-            class ProducesLink extends DefaultTask {
-                @OutputDirectory File outputDirectory // TODO outputfile?
-    
-                @TaskAction execute() {
-                    def link = Paths.get('${link}')
-                    Files.deleteIfExists(link);
-                    Files.createSymbolicLink(link, Paths.get('${target}'));
-                }
-            }
-            
-            task producesLink(type: ProducesLink) {
-                outputDirectory = file '${root.name}'
-            }
-        """
+        buildFile << script(target, link)
 
         when:
         target.createFile()
@@ -110,6 +94,48 @@ class FileCollectionSymlinkIntegrationTest extends AbstractIntegrationSpec {
         run 'producesLink'
         then:
         skipped ':producesLink'
+
+        where:
+        outputType        | script
+        'OutputDirectory' | { targetParam, linkParam -> symbolicLinkOutputDirectory(targetParam, linkParam) }
+        'OutputFile'      | { targetParam, linkParam -> symbolicLinkOutputFile(targetParam, linkParam) }
+    }
+
+    def symbolicLinkOutputDirectory(target, link) {
+        """
+            import java.nio.file.*
+            class ProducesLink extends DefaultTask {
+                @OutputDirectory File outputDirectory 
+    
+                @TaskAction execute() {
+                    def link = Paths.get('${link}')
+                    Files.deleteIfExists(link);
+                    Files.createSymbolicLink(link, Paths.get('${target}'));
+                }
+            }
+            
+            task producesLink(type: ProducesLink) {
+                outputDirectory = file '${link.parentFile}'
+            }
+        """
+    }
+
+    def symbolicLinkOutputFile(target, link) {
+        """
+            import java.nio.file.*
+            class ProducesLink extends DefaultTask {
+                @OutputFile Path outputFile
+    
+                @TaskAction execute() {
+                    Files.deleteIfExists(outputFile);
+                    Files.createSymbolicLink(outputFile, Paths.get('${target}'));
+                }
+            }
+            
+            task producesLink(type: ProducesLink) {
+                outputFile = Paths.get('${link}')
+            }
+        """
     }
 
     @Issue('https://github.com/gradle/gradle/issues/1365')

--- a/subprojects/docs/src/docs/release/notes.md
+++ b/subprojects/docs/src/docs/release/notes.md
@@ -27,6 +27,7 @@ Include only their name, impactful features should be called out separately belo
 [Stephan Windmüller](https://github.com/stovocor),
 [Zemian Deng](https://github.com/zemian),
 [Robin Verduijn](https://github.com/robinverduijn),
+[idlsoft](https://github.com/idlsoft),
 and [Christian Fränkel](https://github.com/fraenkelc).
 
 ## Upgrade Instructions

--- a/subprojects/docs/src/docs/release/notes.md
+++ b/subprojects/docs/src/docs/release/notes.md
@@ -27,7 +27,7 @@ Include only their name, impactful features should be called out separately belo
 [Stephan Windmüller](https://github.com/stovocor),
 [Zemian Deng](https://github.com/zemian),
 [Robin Verduijn](https://github.com/robinverduijn),
-[idlsoft](https://github.com/idlsoft),
+[Sandu Turcan](https://github.com/idlsoft),
 and [Christian Fränkel](https://github.com/fraenkelc).
 
 ## Upgrade Instructions

--- a/subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java
+++ b/subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java
@@ -23,7 +23,6 @@ import com.google.common.collect.Lists;
 import org.gradle.internal.hash.FileHasher;
 import org.gradle.internal.hash.HashCode;
 import org.gradle.internal.snapshot.FileMetadata;
-import org.gradle.internal.snapshot.FileSnapshottingException;
 import org.gradle.internal.snapshot.FileSystemLocationSnapshot;
 import org.gradle.internal.snapshot.MerkleDirectorySnapshotBuilder;
 import org.gradle.internal.snapshot.MissingFileSnapshot;
@@ -190,10 +189,6 @@ public class DirectorySnapshotter {
         public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
             String internedName = intern(file.getFileName().toString());
             if (shouldVisit(file, internedName, false, attrs, builder.getRelativePath())) {
-                if (attrs.isSymbolicLink()) {
-                    // when FileVisitOption.FOLLOW_LINKS, we only get here when link couldn't be followed
-                    throw new FileSnapshottingException(String.format("Could not list contents of '%s'. Couldn't follow symbolic link.", file));
-                }
                 builder.visitFile(snapshotFile(file, internedName, attrs));
             }
             return FileVisitResult.CONTINUE;

--- a/subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/impl/DirectorySnapshotterTest.groovy
+++ b/subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/impl/DirectorySnapshotterTest.groovy
@@ -156,6 +156,26 @@ class DirectorySnapshotterTest extends Specification {
     }
 
     @Requires(TestPrecondition.FILE_PERMISSIONS)
+    def "broken symlinks are snapshotted as missing"() {
+        def rootDir = tmpDir.createDir("root")
+        rootDir.file('brokenSymlink').createLink("linkTarget")
+        assert rootDir.listFiles()*.exists() == [false]
+        def predicate = directoryWalkerPredicate(new PatternSet())
+
+        when:
+        def brokenSymlinkSnapshot = directorySnapshotter.snapshot(rootDir.absolutePath, predicate, new AtomicBoolean(false))
+        then:
+        brokenSymlinkSnapshot.children*.class == [MissingFileSnapshot]
+
+        when:
+        rootDir.file("linkTarget").createFile() // unbreak my heart
+        and:
+        def unbrokenSymlinkSnapshot = directorySnapshotter.snapshot(rootDir.absolutePath, predicate, new AtomicBoolean(false))
+        then:
+        unbrokenSymlinkSnapshot.children*.class == [RegularFileSnapshot, RegularFileSnapshot]
+    }
+
+    @Requires(TestPrecondition.FILE_PERMISSIONS)
     def "unreadable files and directories are snapshotted as missing"() {
         given:
         def rootDir = tmpDir.createDir("root")

--- a/subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/impl/DirectorySnapshotterTest.groovy
+++ b/subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/impl/DirectorySnapshotterTest.groovy
@@ -160,17 +160,15 @@ class DirectorySnapshotterTest extends Specification {
         def rootDir = tmpDir.createDir("root")
         rootDir.file('brokenSymlink').createLink("linkTarget")
         assert rootDir.listFiles()*.exists() == [false]
-        def predicate = directoryWalkerPredicate(new PatternSet())
-
         when:
-        def brokenSymlinkSnapshot = directorySnapshotter.snapshot(rootDir.absolutePath, predicate, new AtomicBoolean(false))
+        def brokenSymlinkSnapshot = directorySnapshotter.snapshot(rootDir.absolutePath, null, new AtomicBoolean(false))
         then:
         brokenSymlinkSnapshot.children*.class == [MissingFileSnapshot]
 
         when:
         rootDir.file("linkTarget").createFile() // unbreak my heart
         and:
-        def unbrokenSymlinkSnapshot = directorySnapshotter.snapshot(rootDir.absolutePath, predicate, new AtomicBoolean(false))
+        def unbrokenSymlinkSnapshot = directorySnapshotter.snapshot(rootDir.absolutePath, null, new AtomicBoolean(false))
         then:
         unbrokenSymlinkSnapshot.children*.class == [RegularFileSnapshot, RegularFileSnapshot]
     }

